File: cv_1/twoDigits.py
import sys

# Digits are encoded between 48 and 57
LOW_MARGIN_ASCII_DIG = 48
TOP_MARGIN_ASCII_DIG = 57

def toASCII(char:str) -> int:
    return ord(char)

def isValidDigitASCII(charASCII:str) -> bool:
    return charASCII >= LOW_MARGIN_ASCII_DIG and charASCII <= TOP_MARGIN_ASCII_DIG

def getLastTwoInputRowDigits() -> int:
    penultimateASCII, ultimateASCII = None, None
    currCharASCII = toASCII(getNextChar())

    while (isValidDigitASCII(currCharASCII)):
        penultimateASCII, ultimateASCII = ultimateASCII, currCharASCII
        currCharASCII = toASCII(getNextChar())
    
    ultimate = ultimateASCII- LOW_MARGIN_ASCII_DIG

    if (not penultimateASCII):
        return ultimate
    
    penultimate = penultimateASCII - LOW_MARGIN_ASCII_DIG

    lastTwoDigits = penultimate * 10 + ultimate
    return lastTwoDigits

def getLastNDigits(number:int, n:int = 1) -> int:
    numberStr = str(number)
    return int(numberStr[-n:])

def getNextChar() -> str:
    nextChar = sys.stdin.read(1)
    return nextChar

def getLastTwoDigitsPower():

    baseLastTwo = getLastTwoInputRowDigits() # Only last two digits matter
    cLastTwo = 1

    cPowerCharASCII = toASCII(getNextChar())
    while (isValidDigitASCII(cPowerCharASCII)):
        cDigit = cPowerCharASCII - LOW_MARGIN_ASCII_DIG

        if (cDigit == 0):
            cLastTwo = getLastNDigits(cLastTwo * cLastTwo, 2)
        elif (cDigit == 1):
            cLastTwo = getLastNDigits(cLastTwo * cLastTwo * baseLastTwo, 2)
        
        cPowerCharASCII = toASCII(getNextChar())

    cLastTwoStr = str(cLastTwo) if len(str(cLastTwo)) >= 2 else "0" + str(cLastTwo)

    print(cLastTwoStr)
    return



getLastTwoDigitsPower()

File: cv_3/twoDigits/twoDigits.c
#include "module.h"
#include <math.h>
#include <stdio.h>
#define LOW_MARGIN_ASCII_DIG 48;

int getLastTwoInputRowDigits(){
    int penultimateASCII = 0, ultimateASCII = 0;

    int cChar = getNextCharASCII();
    while (isDigit(cChar)){
        penultimateASCII = ultimateASCII;
        ultimateASCII = cChar;
        cChar = getNextCharASCII();
    }

    int ultimate = ultimateASCII -  LOW_MARGIN_ASCII_DIG;

    if (!penultimateASCII){
        return ultimate;
    }

    int penultimate = penultimateASCII - LOW_MARGIN_ASCII_DIG;

    return penultimate * 10 + ultimate;
}

int getLastNDigits(int number, int n){
    int base = 10;
    int lastNDigits =  number % (int)pow(base, n);
    return lastNDigits;
}

int main(){
    int baseLastTwo = getLastTwoInputRowDigits();
    
    int cLastTwo = 1;
    int cPowerCharASCII = getNextCharASCII();
    while (isDigit(cPowerCharASCII)){
        int cDigit = cPowerCharASCII - LOW_MARGIN_ASCII_DIG;

        switch (cDigit){
            case 0:
                cLastTwo = getLastNDigits(cLastTwo * cLastTwo, 2);
                break;
            case 1:
                cLastTwo = getLastNDigits(cLastTwo * cLastTwo * baseLastTwo, 2);
                break;
        }

        cPowerCharASCII = getNextCharASCII();
    }

    if (cLastTwo < 10){
        printf("0%d\n", cLastTwo);
        return 0;
    }

    printf("%d\n", cLastTwo);
    return 0;
}